core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/ArrayBindingBasedQueryEvaluationContext.java

```diff
@@ -131,51 +239,81 @@ public BiConsumer<Value, MutableBindingSet> addBinding(String variableName) {
+				true) {
 
 			@Override
 			public void meet(Var node) throws QueryEvaluationException {
 				super.meet(node);
-				varNames.add(node.getName());
+				// We can skip constants that are only used in StatementPatterns since these are never added to the
+				// BindingSet anyway
+				if (!(node.isConstant() && node.getParentNode() instanceof StatementPattern)) {
+					Var replacement = new Var(varNames.computeIfAbsent(node.getName(), k -> k), node.getValue(),
+							node.isAnonymous(), node.isConstant());
+					node.replaceWith(replacement);
+				}
 			}
 
 			@Override
 			public void meet(ProjectionElem node) throws QueryEvaluationException {
-				varNames.add(node.getSourceName());
-				varNames.add(node.getTargetName());
 				super.meet(node);
+				node.setSourceName(varNames.computeIfAbsent(node.getSourceName(), k -> k));
+				node.setTargetName(varNames.computeIfAbsent(node.getTargetName(), k -> k));
+			}
+
+			@Override
+			protected void meetUnaryTupleOperator(UnaryTupleOperator node) throws QueryEvaluationException {
+				if (node instanceof Projection) {
+					node.getArg().visit(this);
+					((Projection) node).getProjectionElemList().visit(this);
+				} else {
+					node.visitChildren(this);
+				}
 			}
 
 			@Override
 			public void meet(MultiProjection node) throws QueryEvaluationException {
-				varNames.addAll(node.getBindingNames());
+				for (String bindingName : node.getBindingNames()) {
+					varNames.computeIfAbsent(bindingName, k -> k);
+				}
 				super.meet(node);
 			}
 
 			@Override
 			public void meet(ZeroLengthPath node) throws QueryEvaluationException {
-				varNames.add(ZeroLengthPathIteration.ANON_SUBJECT_VAR);
-				varNames.add(ZeroLengthPathIteration.ANON_PREDICATE_VAR);
-				varNames.add(ZeroLengthPathIteration.ANON_OBJECT_VAR);
-				varNames.add(ZeroLengthPathIteration.ANON_SEQUENCE_VAR);
+				varNames.computeIfAbsent(ZeroLengthPathIteration.ANON_SUBJECT_VAR, k -> k);
+				varNames.computeIfAbsent(ZeroLengthPathIteration.ANON_PREDICATE_VAR, k -> k);
+				varNames.computeIfAbsent(ZeroLengthPathIteration.ANON_OBJECT_VAR, k -> k);
+				varNames.computeIfAbsent(ZeroLengthPathIteration.ANON_SEQUENCE_VAR, k -> k);
 				super.meet(node);
 			}
 
 			@Override
 			public void meet(ExtensionElem node) throws QueryEvaluationException {
-				varNames.add(node.getName());
+				node.setName(varNames.computeIfAbsent(node.getName(), k -> k));
+				super.meet(node);
+			}
+
+			@Override
+			public void meet(Group node) throws QueryEvaluationException {
+				List<String> collect = node.getGroupBindingNames()
+						.stream()
+						.map(varName -> varNames.computeIfAbsent(varName, k -> k))
+						.collect(Collectors.toList());
+				node.setGroupBindingNames(collect);
 				super.meet(node);
 			}
 		};
 		node.visit(queryModelVisitorBase);
-		String[] varNamesArr = varNames.toArray(new String[0]);
-		return varNamesArr;
+		return varNames.keySet().toArray(new String[0]);
 	}
 }
```

Comment:
This code deduplicates var name strings so that vars with the same name use the same string objects. It isn't meant as a guarantee, but rather as an aid to allow some optimizations. Even though we still need to call `.equals(...)` to be certain, we can now first use `==`.

Reply:
Excellent, this was on my todo list to figure out.